nidx/src/maintenance/scheduler.rs

```diff
@@ -158,23 +169,112 @@ pub async fn purge_deletions(meta: &NidxMetadata, consumer: &mut PullConsumer) -
     Ok(())
 }
 
-pub async fn schedule_merges(meta: &NidxMetadata, consumer: &mut PullConsumer) -> anyhow::Result<()> {
-    let oldest_confirmed_seq = consumer.info().await?.ack_floor.stream_sequence;
-    let oldest_pending_seq = oldest_confirmed_seq + 1;
-
-    // Enqueue merges. Right now, merge everything that we can
+/// Enqueue merge jobs for segments older than `before_seq` that aren't already
+/// scheduled or marked to delete. Right now we merge everything that we can
+async fn schedule_merges(meta: &NidxMetadata, before_seq: Seq) -> anyhow::Result<()> {
     // TODO: better merge algorithm
-    let merges = sqlx::query_scalar!(
-        r#"SELECT array_agg(id) AS "segment_ids!" FROM segments WHERE delete_at IS NULL AND merge_job_id IS NULL AND seq < $1 GROUP BY index_id"#,
-        oldest_pending_seq as i64
+    let merges = sqlx::query!(
+        r#"SELECT index_id, array_agg(id) AS "segment_ids!: Vec<SegmentId>" FROM segments WHERE delete_at IS NULL AND merge_job_id IS NULL AND seq < $1 GROUP BY index_id"#,
+        i64::from(before_seq),
     )
     .fetch_all(&meta.pool)
     .await?;
     for m in merges {
-        if m.len() > 3 {
-            MergeJob::create(meta, &m, oldest_confirmed_seq as i64).await?;
+        if m.segment_ids.len() > 3 {
+            MergeJob::create(meta, m.index_id.into(), &m.segment_ids, before_seq).await?;
         }
     }
 
     Ok(())
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    mod merge_scheduling {
+        use std::collections::HashSet;
+
+        use nidx_types::Seq;
+        use uuid::Uuid;
+
+        use super::*;
+
+        use crate::metadata::{Index, IndexId, IndexKind, NidxMetadata, Shard};
+
+        #[sqlx::test]
+        async fn test_schedule_merges_for_shard_with_single_index(pool: sqlx::PgPool) {
+            let meta = NidxMetadata::new_with_pool(pool).await.unwrap();
+            let kbid = Uuid::new_v4();
+            let shard = Shard::create(&meta.pool, kbid).await.unwrap();
+            let index = Index::create(&meta.pool, shard.id, IndexKind::Vector, Some("multilingual")).await.unwrap();
+            let mut seq: i64 = 0;
+
+            for _ in 0..10 {
+                let segment = Segment::create(&meta.pool, index.id, Seq::from(seq)).await.unwrap();
+                segment.mark_ready(&meta.pool, 50, 1000).await.unwrap();
+                seq += 1;
+            }
+
+            // creation of shards/indexes/segments don't trigger any merge job
+            assert!(MergeJob::take(&meta.pool).await.unwrap().is_none());
+
+            schedule_merges(&meta, Seq::from(100)).await.unwrap();
+
+            // one job has been scheduled for the index
+            let mut jobs = vec![];
+            while let Some(job) = MergeJob::take(&meta.pool).await.unwrap() {
+                jobs.push(job)
+            }
+            assert_eq!(jobs.len(), 1);
+            assert_eq!(IndexId::from(jobs[0].index_id), index.id);
```

Comment:
Maybe add a couple of asserts:
- The merge job links to all 10 segments
- The seq from the merge job is the same as in the scheduler.